### ISSUE TITLE
[BUG] Certificates do not renew successfully using crc v1.10

### ISSUE THREAD

**Issue by btannous** [code-ready/crc#1242]
### General information

  * OS: Linux (fedora 31)
  * Hypervisor: KVM
  * Did you run `crc setup` before starting it (Yes/No)? Yes (always)
  * Running CRC on: Desktop, 8core, 32g memory, nothing else running (accessing remotely)

## CRC version
```bash
❯ ./crc version
crc version: 1.10.0+9025021
OpenShift version: 4.4.3 (embedded in binary)
```

## Host Operating System
```bash
❯ cat /etc/os-release
NAME=Fedora
VERSION="31 (Server Edition)"
ID=fedora
VERSION_ID=31
VERSION_CODENAME=""
PLATFORM_ID="platform:f31"
PRETTY_NAME="Fedora 31 (Server Edition)"
ANSI_COLOR="0;34"
LOGO=fedora-logo-icon
CPE_NAME="cpe:/o:fedoraproject:fedora:31"
HOME_URL="https://fedoraproject.org/"
DOCUMENTATION_URL="https://docs.fedoraproject.org/en-US/fedora/f31/system-administrators-guide/"
SUPPORT_URL="https://fedoraproject.org/wiki/Communicating_and_getting_help"
BUG_REPORT_URL="https://bugzilla.redhat.com/"
REDHAT_BUGZILLA_PRODUCT="Fedora"
REDHAT_BUGZILLA_PRODUCT_VERSION=31
REDHAT_SUPPORT_PRODUCT="Fedora"
REDHAT_SUPPORT_PRODUCT_VERSION=31
PRIVACY_POLICY_URL="https://fedoraproject.org/wiki/Legal:PrivacyPolicy"
VARIANT="Server Edition"
VARIANT_ID=server
```

### Steps to reproduce

- blocked UDP port 123 out of my local gateway
- sudo date -s "9 JULY 2020 11:11:11"
- downloaded/extracted crc-1.10
- crc delete <just in case i had an old one>
- crc setup
- crc start -c 8 -m 25000 ...

### Expected
Certificates get renewed and cluster comes up.

### Actual
Certificates do not get renewed.

### Logs
https://gist.github.com/dea1a60dddaf7becd05200d0a2eed3e5

https://gist.github.com/btannous/1ebb796dbb747a2f18294403aea6c15d

**Comment by gbraad** [comment]
We are currently investigating this ...

**Comment by cfergeau** [comment]
> - blocked UDP port 123 out of my local gateway

Turns out this can be done directly with the crc VM on linux
```
$ cat drop-ntp.xml
<filter name='drop-ntp' chain='ipv4'>
  <rule action='drop' direction='out' >
    <ip protocol='udp' dstportstart='123'/>
  </rule>
  <rule action='drop' direction='in' priority='100'>
    <ip protocol='udp' srcportstart='123'/>
  </rule>
</filter>

$ virsh -c qemu:///system net nwfilter-define drop-ntp.xml
$ virsh -c qemu:///system edit crc
```
Then, in crc XML definitions, a `<filterref>` element needs to be added:
```
<domain>
  ...
  <devices>
  ...
    <interface type='network'>
      ...
      <filterref filter='drop-ntp'/>
    </interface>
```

**Comment by BenHall** [comment]
The certificates expired this morning and it looks like this has had an impact. Any suggestions?

**Comment by btannous** [comment]
https://docs.openshift.com/container-platform/4.4/backup_and_restore/disaster_recovery/scenario-3-expired-certs.html
I saw this, not sure if the process changed from pre 4.4 / 4.3 times?

**Comment by btannous** [comment]
https://gist.github.com/dea1a60dddaf7becd05200d0a2eed3e5

**Comment by btannous** [comment]
Is this the issue?
https://github.com/code-ready/snc/blob/35cb5b98bd6575d7e553e23ba337f9e59812fac6/createdisk.sh#L277-L278

```
❯ pwd
/home/b/.crc/machines/crc

❯ ls
config.json  crc  id_rsa  id_rsa.pub  kubeconfig

❯ oc adm release info --config=config.json --image-for=cluster-kube-apiserver-operator
Flag --config has been deprecated, use --kubeconfig instead
error: info expects one argument, or a connection to an OpenShift 4.x server: invalid configuration: Missing or incomplete configuration info.  Please login or point to an existing, complete config file:

  1. Via the command-line flag --kubeconfig
  2. Via the KUBECONFIG environment variable
  3. In your home directory as ~/.kube/config

To view or setup config directly use the 'config' command.
❯ oc adm release info --registry-config=config.json --image-for=cluster-kube-apiserver-operator
quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:1716497ba22b107f22c4d80d7c1e15424aa171f0592dc764d39c0e54c7f5225e
```

**Comment by btannous** [comment]
seems possible:
https://docs.openshift.com/container-platform/4.4/release_notes/ocp-4-4-release-notes.html#ocp-4-4-oc-config-flag

**Comment by phantomjinx** [comment]
@btannous 
Have the same issue and been fighting it all today ...
Bit of debugging shows ...
1. The directory `/etc/kubernetes/static-pod-resources/recovery-kube-apiserver-pod` is created during the start of the recovery pod [here](https://github.com/code-ready/crc/blob/v1.10.0/pkg/crc/machine/cert-renewal.go#L143)
2. The pod then generates the new certificates [here](https://github.com/code-ready/crc/blob/v1.10.0/pkg/crc/machine/cert-renewal.go#L157) 
3. The pod is deferred for removal (so on exit), including its config [here](https://github.com/code-ready/crc/blob/v1.10.0/pkg/crc/machine/cert-renewal.go#L148)

Therefore, once the recovery pod is destroyed the config directory is removed as well.

It is possible to run the command, manually, that created the pod again but the new certificates generated in the `admin.kubeconfig` are not signed by the correct CA so running any `oc` command after just errors out with `unknown certificate-authority errors`.

**Comment by gbraad** [comment]
It seems version 1.8.0 does not have this issue: https://github.com/code-ready/crc/issues/1271#issuecomment-637041784 so it is related to the latest renew implementation. See: https://github.com/code-ready/crc/issues/1271#issuecomment-637041784

**Comment by preinking** [comment]
I can confirm that this issue does not exist in 1.9.0 because this deploys fine from scratch  on my environment. When performing the same using version 1.10.0 crc start fails consistently.

**Logs**
[https://github.com/code-ready/crc/files/4717841/logs.txt](https://github.com/code-ready/crc/files/4717841/logs.txt)

**Comment by rscher** [comment]
same for me … 
ERRO Failed to renew TLS certificates: ssh command error:
command : sudo podman run -it --rm --network=host -v /etc/kubernetes/:/etc/kubernetes/:Z --entrypoint=/usr/bin/cluster-kube-apiserver-operator 'openshift/cert-recovery' regenerate-certificates
err     : exit status 1
output  : I0604 14:57:23.792276       1 certrotationcontroller.go:505] Waiting for CertRotation
I0604 14:57:23.893887       1 client_cert_rotation_controller.go:121] Waiting for CertRotationController - "AggregatorProxyClientCert"
I0604 14:57:23.894033       1 client_cert_rotation_controller.go:128] Finished waiting for CertRotationController - "AggregatorProxyClientCert"
I0604 14:57:23.894038       1 client_cert_rotation_controller.go:121] Waiting for CertRotationController - "KubeAPIServerToKubeletClientCert"
I0604 14:57:23.894042       1 client_cert_rotation_controller.go:128] Finished waiting for CertRotationController - "KubeAPIServerToKubeletClientCert"
I0604 14:57:23.894045       1 client_cert_rotation_controller.go:121] Waiting for CertRotationController - "LocalhostServing"
I0604 14:57:23.894048       1 client_cert_rotation_controller.go:128] Finished waiting for CertRotationController - "LocalhostServing"
I0604 14:57:23.894051       1 client_cert_rotation_controller.go:121] Waiting for CertRotationController - "ServiceNetworkServing"
I0604 14:57:23.894054       1 client_cert_rotation_controller.go:128] Finished waiting for CertRotationController - "ServiceNetworkServing"
I0604 14:57:23.894057       1 client_cert_rotation_controller.go:121] Waiting for CertRotationController - "ExternalLoadBalancerServing"
I0604 14:57:23.894060       1 client_cert_rotation_controller.go:128] Finished waiting for CertRotationController - "ExternalLoadBalancerServing"
I0604 14:57:23.894121       1 client_cert_rotation_controller.go:121] Waiting for CertRotationController - "InternalLoadBalancerServing"
I0604 14:57:23.894144       1 client_cert_rotation_controller.go:128] Finished waiting for CertRotationController - "InternalLoadBalancerServing"
I0604 14:57:23.894161       1 client_cert_rotation_controller.go:121] Waiting for CertRotationController - "LocalhostRecoveryServing"
I0604 14:57:23.894167       1 client_cert_rotation_controller.go:128] Finished waiting for CertRotationController - "LocalhostRecoveryServing"
I0604 14:57:23.894170       1 client_cert_rotation_controller.go:121] Waiting for CertRotationController - "KubeControllerManagerClient"
I0604 14:57:23.894301       1 client_cert_rotation_controller.go:128] Finished waiting for CertRotationController - "KubeControllerManagerClient"
I0604 14:57:23.894305       1 client_cert_rotation_controller.go:121] Waiting for CertRotationController - "KubeSchedulerClient"
I0604 14:57:23.894309       1 client_cert_rotation_controller.go:128] Finished waiting for CertRotationController - "KubeSchedulerClient"
I0604 14:57:23.894339       1 certrotationcontroller.go:527] Finished waiting for CertRotation
I0604 14:57:23.894343       1 kubecontrollermanagercertrotation.go:84] Waiting for CertRotation
I0604 14:57:23.894346       1 client_cert_rotation_controller.go:121] Waiting for CertRotationController - "CSRSigningCert"
I0604 14:57:23.894365       1 client_cert_rotation_controller.go:128] Finished waiting for CertRotationController - "CSRSigningCert"
I0604 14:57:23.894367       1 kubecontrollermanagercertrotation.go:90] Finished waiting for CertRotation
I0604 14:57:23.894421       1 regenerate_certificates.go:197] Refreshing certificates.
I0604 14:57:23.899797       1 event.go:281] Event(v1.ObjectReference{Kind:"namespace", Namespace:"openshift-kube-apiserver-operator", Name:"openshift-kube-apiserver-operator", UID:"", APIVersion:"v1", ResourceVersion:"", FieldPath:""}): type: 'Normal' reason: 'SignerUpdateRequired' "csr-signer-signer" in "openshift-kube-controller-manager-operator" requires a new signing cert/key pair: past its refresh time 2020-06-01 05:40:48 +0000 UTC
I0604 14:57:24.301756       1 event.go:281] Event(v1.ObjectReference{Kind:"namespace", Namespace:"openshift-kube-apiserver-operator", Name:"openshift-kube-apiserver-operator", UID:"", APIVersion:"v1", ResourceVersion:"", FieldPath:""}): type: 'Normal' reason: 'CABundleUpdateRequired' "csr-controller-signer-ca" in "openshift-kube-controller-manager-operator" requires a new cert
I0604 14:57:24.301841       1 event.go:281] Event(v1.ObjectReference{Kind:"namespace", Namespace:"openshift-kube-apiserver-operator", Name:"openshift-kube-apiserver-operator", UID:"", APIVersion:"v1", ResourceVersion:"", FieldPath:""}): type: 'Normal' reason: 'SecretUpdated' Updated Secret/csr-signer-signer -n openshift-kube-controller-manager-operator because it changed
Operation cannot be fulfilled on configmaps "csr-controller-signer-ca": the object has been modified; please apply your changes to the latest version and try again
I0604 14:57:24.597475       1 event.go:281] Event(v1.ObjectReference{Kind:"namespace", Namespace:"openshift-kube-apiserver-operator", Name:"openshift-kube-apiserver-operator", UID:"", APIVersion:"v1", ResourceVersion:"", FieldPath:""}): type: 'Warning' reason: 'ConfigMapUpdateFailed' Failed to update ConfigMap/csr-controller-signer-ca -n openshift-kube-controller-manager-operat r: Operation cannot be fulfilled on configmaps "csr-controller-signer-ca": the object has been modified; please apply your changes to the latest version and try again
=========
$ crc version
crc version: 1.10.0+9025021
OpenShift version: 4.4.3 (embedded in binary)

**Comment by rscher** [comment]
same here, cant start after applying crc 1.10.0 on Mac-

ERRO Failed to renew TLS certificates: ssh command error:
command : sudo podman run -it --rm --network=host -v /etc/kubernetes/:/etc/kubernetes/:Z --entrypoint=/usr/bin/cluster-kube-apiserver-operator 'openshift/cert-recovery' regenerate-certificates

$ crc version
crc version: 1.10.0+9025021
OpenShift version: 4.4.3 (embedded in binary)
===
Is there a patch ? I'm dead in the water

**Comment by btannous** [comment]
Just for documentation on this issue, I tested 1.8-1.11.
Looks like certs renew properly on all CRC version `<1.10` and they do not renew on `>=1.10`.

**Comment by rscher** [comment]
> Just for documentation on this issue, I tested 1.8-1.11.
> Looks like certs renew properly on all CRC version `<1.10` and they do not renew on `>=1.10`.

Where can I download the older versions crc 1.9  or 1.8  ?

**Comment by gbraad** [comment]
https://mirror.openshift.com/pub/openshift-v4/clients/crc/

**Comment by tadayosi** [comment]
For the time being, upgrading crc to latest 1.11.0 solves temporarily (i.e. postpones for another 30 days) the issue.

**Comment by btannous** [comment]
As commented in the PR https://github.com/code-ready/crc/pull/1292 , I validated that the fix works on my end, so once merged in and a new build comes out this should be *likely* closed.

**Comment by btannous** [comment]
@praveenkumar @cfergeau wooo! thank you all. 

@BenHall certs should be good now (when a new build comes i’d assume) so unblocked :)!

**Comment by BenHall** [comment]
@btannous Thanks! I applied the changes this morning and will test properly when the certs expire at the end of the month.
